pkg/gds/secret.go

```diff
@@ -28,108 +27,180 @@ func CreateToken(length int) string {
+	sctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
+
+	// Build the request.
+	req := &secretmanagerpb.AddSecretVersionRequest{
+		Parent: secretPath,
+		Payload: &secretmanagerpb.SecretPayload{
+			Data: payload,
+		},
 	}
 
-	sig, err := createHMAC(key, ciphertext)
+	// Call the API. Note: We don't actually need the result that comes back from the API call
+	// and not accessing it directly (e.g. logging plaintext, etc) provides added security
+	_, err := sm.client.AddSecretVersion(sctx, req)
 	if err != nil {
-		return nil, nil, err
+		// If the API call is malformed, it will hang until the internal context times out
+		if errors.Is(err, context.DeadlineExceeded) {
+			return err
+		}
+
+		// If the secret does not exist (e.g. has been deleted or hasn't been created yet)
+		// we'll get a Not Found error
+		serr, ok := status.FromError(err)
+		if ok && serr.Code() == codes.NotFound {
+			return err
+		}
+
+		// If the error is something else, something went wrong.
+		return errors.New("unknown error: unable to create secret version")
 	}
 
-	// Concatenate the ciphertext and the nonce to facilitate decryption
-	ciphertext = append(ciphertext, nonce...)
-	return ciphertext, sig, nil
+	return nil
 }
 
-// Decrypt the ciphertext with the server's secret key and verify the HMAC.
-func (s *Server) Decrypt(ciphertext, signature []byte) (plaintext string, err error) {
-	if len(ciphertext) == 0 {
-		return "", errors.New("empty cipher text")
-	}
+// GetLatestVersion returns the payload for the latest version of the given secret,
+// if one exists, else an error.
+func (sm *SecretManager) GetLatestVersion(ctx context.Context, secret string) ([]byte, error) {
 
-	// Create a 32 byte signature of the key
-	hash := sha256.New()
-	hash.Write([]byte(s.conf.SecretKey))
-	key := hash.Sum(nil)
+	versionPath := fmt.Sprintf("%s/secrets/%s-%s/versions/latest", sm.parent, sm.requestId, secret)
 
-	// Separate the data from the nonce
-	data := ciphertext[:len(ciphertext)-nonceSize]
-	nonce := ciphertext[len(ciphertext)-nonceSize:]
+	// Create an internal context, since a failed API call will result in infinite hang
+	sctx, cancel := context.WithTimeout(ctx, 30*time.Second)
+	defer cancel()
 
-	// Validate HMAC signature
-	if err = validateHMAC(key, data, signature); err != nil {
-		return "", err
+	// Build the request.
+	req := &secretmanagerpb.AccessSecretVersionRequest{
+		Name: versionPath,
 	}
 
-	block, err := aes.NewCipher(key)
+	// Call the API.
+	result, err := sm.client.AccessSecretVersion(sctx, req)
 	if err != nil {
-		return "", err
-	}
+		// If the API call is malformed, it will hang until the internal context times out
+		if errors.Is(err, context.DeadlineExceeded) {
+			return nil, err
+		}
 
-	aesgcm, err := cipher.NewGCM(block)
-	if err != nil {
-		return "", err
+		// If the error is something else, something went wrong.
+		return nil, errors.New("unknown error: unable to access latest secret version")
```

Comment:
Could we also make this `return nil, fmt.Errorf("unable to access latest secret version: %s", err)` - thank you!